samples/gitchurn/src/gitChurnMain/kotlin/GitChurn.kt

```diff
@@ -35,33 +35,44 @@ fun main(args: Array<String>) {
 private fun calculateChurn(workDir: String, limit: Int) {
     println("Opening…")
     val repository = git.repository(workDir)
-    val map = mutableMapOf<String, Int>()
+    data class WhoAndWhere(val authorEmail: String = "", val filePath: String = "")
+    val modificationsByAuthor = mutableMapOf<WhoAndWhere, Int>()
     var count = 0
-    val commits = repository.commits()
-    val limited = commits.take(limit)
+    val commits = repository.myCommits()
+    val commitsLimit = commits.take(limit)
     println("Calculating…")
-    limited.forEach { commit ->
+    commitsLimit.forEach { commit ->
         if (count % 100 == 0)
-            println("Commit #$count [${commit.time.format()}]: ${commit.summary}")
+            println("Commit #$count [${commit.time.format()}] by ${commit.author.name!!.toKString()}: ${commit.summary}")
 
         commit.parents.forEach { parent ->
             val diff = commit.tree.diff(parent.tree)
             diff.deltas().forEach { delta ->
-                val path = delta.newPath
-                val n = map[path] ?: 0
-                map.put(path, n + 1)
+                val path = if (delta.newPath == delta.newPath.substringBefore('/'))
+                    delta.newPath
+                else
+                    delta.newPath.substringBefore('/') + "/"
+                val n = modificationsByAuthor[WhoAndWhere(parent.author.email!!.toKString().toLowerCase(), path)] ?: 0
+                modificationsByAuthor[WhoAndWhere(parent.author.email!!.toKString().toLowerCase(), path)] = n + 1
             }
             diff.close()
             parent.close()
         }
         commit.close()
         count++
     }
-    println("Report:")
-    map.toList().sortedByDescending { it.second }.take(10).forEach {
-        println("File: ${it.first}")
-        println("      ${it.second}")
-        println()
+    println("Named Report:")
+    modificationsByAuthor.toList().groupBy { it.first.authorEmail }.values.forEach {
```

Comment:
Lack of named variables and properties (other than almost meaningless `it`, `.first` and `.second`) makes this code complicated.

Reply:
I don't know, how to deal with groupBy. To use it I got to make List, and then my data comes to be paired, which leads to all this `.first-.second` thing. Is there something I misunderstand, or it can't be solved this way?

Reply:
Consider using [destructuring declarations in lambdas](https://kotlinlang.org/docs/reference/multi-declarations.html#destructuring-in-lambdas-since-11).

Try giving proper names to entities, e.g. to `forEach {` lambda parameters. If you find it extremely difficult, then this must be a sign that your data is too complicated and should be simplified.